--- a/drivers/tty/serial/8250/8250_dw.c	2016-03-18 02:09:52.000000000 +0800
+++ b/drivers/tty/serial/8250/8250_dw.c	2021-05-25 13:41:47.000000000 +0800
@@ -33,6 +33,14 @@
 
 #include "8250.h"
 
+#if defined(CONFIG_LORAWAN_OPENWRT)
+#include <linux/serial.h>
+#include <linux/gpio.h>
+#include <linux/of_gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <linux/delay.h>
+#endif
+
 /* Offsets for the DesignWare specific registers */
 #define DW_UART_USR	0x1f /* UART Status Register */
 #define DW_UART_CPR	0xf4 /* Component Parameter Register */
@@ -345,6 +353,13 @@
 	return 0;
 }
 
+#if defined(CONFIG_LORAWAN_OPENWRT)
+extern int dw8250_config_rs485(struct uart_port *port, 
+	struct serial_rs485 *rs485);
+extern int dw8250_probe_rs485(struct uart_8250_port *up,
+	struct device_node *np);
+#endif
+
 static int dw8250_probe(struct platform_device *pdev)
 {
 	struct uart_8250_port uart = {};
@@ -439,6 +454,13 @@
 		err = data->line;
 		goto err_reset;
 	}
+	
+#if defined(CONFIG_LORAWAN_OPENWRT)
+    err = dw8250_probe_rs485(&uart, pdev->dev.of_node);
+    if (err < 0)
+        return -ENOMEM;
+    uart.rs485_config = dw8250_config_rs485;
+#endif
 
 	platform_set_drvdata(pdev, data);
 
